compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -115,12 +115,18 @@ struct DistributeConstants final : OpDistributionPattern<arith::ConstantOp> {
   }
 };
 
-template <typename OpTy>
-struct DistributeElementwise final : OpDistributionPattern<OpTy> {
-  using OpDistributionPattern<OpTy>::OpDistributionPattern;
+struct DistributeElementwise final
+    : OpTraitDistributionPattern<OpTrait::Elementwise> {
+  using OpTraitDistributionPattern::OpTraitDistributionPattern;
 
-  LogicalResult matchAndRewrite(OpTy op, DistributionSignature &signature,
+  LogicalResult matchAndRewrite(Operation *op, DistributionSignature &signature,
                                 PatternRewriter &rewriter) const override {
+    // Check if this operation has elementwise mappable traits. This is
+    // more restricted than only having elementwise trait.
+    if (!OpTrait::hasElementwiseMappableTraits(op)) {
+      return failure();
+    }
+
```

Comment:
I can't comment on parts that aren't changing, but this should just inherit all attributes rather than just fast math flags.
```
  return builder.create(loc, op->getName().getIdentifier(), operands,
                        resultTypes, op->getAttrs());
```

Reply:
I'm not sure if that's right. VectorDistribution uses attributes to understand what still needs to be distributed. Also, I don't think there is a contract that attributes should be preserved (other than the ones that mean something like FastMathAttrs)

Reply:
Happy to look at this later, but for now, it will need some reworking on things if we retain all attributes.

Reply:
ok no worries if it's hard to handle then.